crates/subspace-networking/src/node_runner.rs

```diff
@@ -237,6 +239,40 @@ where
         }
     }
 
+    /// Bootstraps Kademlia network
+    pub async fn bootstrap(&mut self) {
```

Comment:
This will work as expected most of the time, but not all the time.

`NodeRunner::run()` can be called, driven for some time and then dropped. When it it called again, it could be that you're triggering second, third and so on bootstrapping process.

`Command::Bootstrap` should be sent once in constructor with receiver stored in `NodeRunner`. Then here you need to check if it is still present and only run this function if it is.

Just remember that any async function can be interrupted at any time.

Also I noticed `handle_command` is async, which is not desired and not necessary if you change `remove_all_known_peer_addresses` to not be async anymore. Otherwise you have even more edge cases to potentially deal with.

Reply:
I'm not sure whether we have this case in practice but this feature could be achieved using just a bool.

Reply:
No, it is not. There is absolutely no guarantee that bootstrapping has finished last time you called it. There is a guarantee that it has started bootstrapping process, but not more than that. It will not prevent other commands from being processed in the meantime.